internal/catalog/internal/controllers/endpoints/controller.go

```diff
@@ -0,0 +1,358 @@
+	status, found := workload.Status[workloadhealth.StatusKey]
+	if !found {
+		return pbcatalog.Health_HEALTH_CRITICAL
+	}
+
+	for _, condition := range status.Conditions {
+		if condition.Type == workloadhealth.StatusConditionHealthy {
+			raw, found := pbcatalog.Health_value[condition.Reason]
+			if found {
+				return pbcatalog.Health(raw)
+			}
+			return pbcatalog.Health_HEALTH_CRITICAL
+		}
+	}
+	return pbcatalog.Health_HEALTH_CRITICAL
+}
+
+// serviceUnderManagement detects whether this service should have its
+// endpoints automatically managed by the controller
+func serviceUnderManagement(svc *pbcatalog.Service) bool {
+	sel := svc.GetWorkloads()
+	if sel == nil {
+		// The selector wasn't present at all. Therefore this service is not under
+		// automatic endpoint management.
+		return false
+	}
+
+	if len(sel.Names) < 1 && len(sel.Prefixes) < 1 {
+		// The selector was set in the request but the list of workload names
+		// and prefixes were both empty. Therefore this service is not under
+		// automatic endpoint management
+		return false
+	}
+
+	// Some workload selection criteria exists, so this service is consider
+	// under automatic endpoint management.
+	return true
+}
+
+// workloadsToEndpoints will translate the Workload resources into a ServiceEndpoints resource
+func workloadsToEndpoints(svc *pbcatalog.Service, workloads []*workloadData) *pbcatalog.ServiceEndpoints {
+	var endpoints []*pbcatalog.Endpoint
+
+	for _, workload := range workloads {
+		endpoint := workloadToEndpoint(svc, workload)
+		if endpoint != nil {
+			endpoints = append(endpoints, endpoint)
+		}
+	}
+
+	return &pbcatalog.ServiceEndpoints{
+		Endpoints: endpoints,
+	}
+}
+
+// workloadToEndpoint will convert a workload resource into a singular Endpoint to be
+// put within a ServiceEndpoints resource.
+//
+// The conversion process involves parsing the workloads health and filtering its
+// addresses and ports down to just what the service wants to consume.
+//
+// Determining the workloads health requires the workload-health controller to already
+// have reconciled the workloads health and stored it within the resources Status field.
+// Any unreconciled workload health will be represented in the ServiceEndpoints with
+// the ANY health status.
+func workloadToEndpoint(svc *pbcatalog.Service, data *workloadData) *pbcatalog.Endpoint {
+	health := determineWorkloadHealth(data.resource)
+
+	endpointPorts := make(map[string]*pbcatalog.WorkloadPort)
+
+	// Create the endpoints filtered ports map. Only workload ports specified in
+	// one of the services ports are included. Ports with a protocol mismatch
+	// between the service and workload will be excluded as well.
+	for _, svcPort := range svc.Ports {
+		workloadPort, found := data.workload.Ports[svcPort.TargetPort]
+		if !found {
+			// this workload doesn't have this port so ignore it
+			continue
+		}
+
+		if workloadPort.Protocol != svcPort.Protocol {
+			// workload port mismatch - ignore it
```

Comment:
Where do you think we should report that. Presumably on the service or service endpoints?

Reply:
I think on the service because that's where user provided invalid config.

Reply:
Is it the service thats invalid though?

If 9/10 workloads have that port with the correct protocol but 1/10 has the mismatch we could argue that its the workload thats the problem.

I think adding the status to the workload isn't right as the workload shouldn't need to care about service concerns in theory.

So adding it to the service could be fine but I could also see this being a status on the endpoints.

Reply:
That's fair. I still think in the case you mentioned it could also be the service that's misconfigured because maybe you've selected a workload that shouldn't be part of the service. Ultimately, we don't know what the user's intention was and so saying that something is misconfigured on the service makes sense to me even if it's the workload port that needs to change in the end. 

We could maybe add it to both. My main thought is that adding status to objects that user interacts with has a higher chance of them seeing the error and taking an action. I do agree that adding it to the workload doesn't feel right.

Reply:
This was getting complex enough that I thought it would be best to leave it to a future PR.